src/screens/DiscoveryScreenV2/fragment/TopicFragment.js

```diff
@@ -66,39 +78,74 @@ const TopicFragment = ({
     parseToken();
   }, []);
 
-  const handleFollow = async (from, willFollow, item, index) => {
+  const exhangeFollower = (newTopicLists, willFollow, topicId) => {
+    const indexTopic = newTopicLists.findIndex((item) => item.topic_id === topicId);
+    newTopicLists[indexTopic].following = !!willFollow;
+    newTopicLists[indexTopic].user_id_follower = myId;
+    return newTopicLists[indexTopic];
+  };
+
+  const mapTopic = (newTopic) => {
+    return discovery.initialTopics.map((topic) => {
+      if (topic.topic_id === newTopic.topic_id) {
+        return newTopic;
+      }
+      return topic;
+    });
+  };
+
+  const handleTopic = (from, willFollow, item, index) => {
     if (from === FROM_FOLLOWED_TOPIC_INITIAL) {
       const newFollowedTopics = [...topics];
-      newFollowedTopics[index].following = !!willFollow;
-      newFollowedTopics[index].user_id_follower = myId;
-      DiscoveryAction.setDiscoveryInitialTopics(newFollowedTopics, discoveryDispatch);
+      const newTopic = exhangeFollower(newFollowedTopics, willFollow, item.topic_id);
+
+      DiscoveryAction.setDiscoveryInitialTopics(mapTopic(newTopic), discoveryDispatch);
     }
 
     if (from === FROM_UNFOLLOWED_TOPIC_INITIAL) {
       const newFollowedTopics = [...topics];
-      newFollowedTopics[index].following = !!willFollow;
-      newFollowedTopics[index].user_id_follower = myId;
+      const newTopic = exhangeFollower(newFollowedTopics, willFollow, item.topic_id);
 
-      DiscoveryAction.setDiscoveryInitialTopics(newFollowedTopics, discoveryDispatch);
+      DiscoveryAction.setDiscoveryInitialTopics(mapTopic(newTopic), discoveryDispatch);
     }
 
     if (from === FROM_FOLLOWED_TOPIC) {
-      const newFollowedTopics = [...followedTopic];
-      newFollowedTopics[index].following = !!willFollow;
-      newFollowedTopics[index].user_id_follower = myId;
+      const newFollowedTopics = [...newMapFollowedTopics];
+      const newTopic = exhangeFollower(newFollowedTopics, willFollow, item.topic_id);
 
-      setFollowedTopic(newFollowedTopics);
+      setFollowedTopic(
+        followedTopic.map((topic) => {
+          if (topic.topic_id === newTopic.topic_id) {
+            return newTopic;
+          }
+          return topic;
+        })
+      );
     }
 
     if (from === FROM_UNFOLLOWED_TOPIC) {
-      const newUnFollowedTopic = [...unfollowedTopic];
-      newUnFollowedTopic[index].following = !!willFollow;
-      newUnFollowedTopic[index].user_id_follower = myId;
+      const newUnFollowedTopic = [...newMapUnfollowedTopics];
+      const newTopic = exhangeFollower(newUnFollowedTopic, willFollow, item.topic_id);
 
-      setUnfollowedTopic(newUnFollowedTopic);
+      setUnfollowedTopic(
+        unfollowedTopic.map((topic) => {
+          if (topic.topic_id === newTopic.topic_id) {
+            return newTopic;
+          }
+          return topic;
+        })
+      );
     }
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `handleTopic` function has repeated code blocks for different conditions. This is another violation of the DRY principle. You can create a helper function to handle the common logic.

```diff
+  const updateTopicList = (from, topicList, willFollow, item) => {
+    const newTopicList = [...topicList];
+    const newTopic = exhangeFollower(newTopicList, willFollow, item.topic_id);
+
+    if (from === FROM_FOLLOWED_TOPIC_INITIAL || from === FROM_UNFOLLOWED_TOPIC_INITIAL) {
+      DiscoveryAction.setDiscoveryInitialTopics(mapTopic(newTopic), discoveryDispatch);
+    } else if (from === FROM_FOLLOWED_TOPIC) {
+      setFollowedTopic(
+        followedTopic.map((topic) => {
+          if (topic.topic_id === newTopic.topic_id) {
+            return newTopic;
+          }
+          return topic;
+        })
+      );
+    } else if (from === FROM_UNFOLLOWED_TOPIC) {
+      setUnfollowedTopic(
+        unfollowedTopic.map((topic) => {
+          if (topic.topic_id === newTopic.topic_id) {
+            return newTopic;
+          }
+          return topic;
+        })
+      );
+    }
+  };

-  const handleTopic = (from, willFollow, item, index) => {
-     if (from === FROM_FOLLOWED_TOPIC_INITIAL) {
-       const newFollowedTopics = [...topics];
-       const newTopic = exhangeFollower(newFollowedTopics, willFollow, item.topic_id);
-       DiscoveryAction.setDiscoveryInitialTopics(mapTopic(newTopic), discoveryDispatch);
-     }
-     if (from === FROM_UNFOLLOWED_TOPIC_INITIAL) {
-       const newFollowedTopics = [...topics];
-       const newTopic = exhangeFollower(newFollowedTopics, willFollow, item.topic_id);
-       DiscoveryAction.setDiscoveryInitialTopics(mapTopic(newTopic), discoveryDispatch);
-     }
-     if (from === FROM_FOLLOWED_TOPIC) {
-       const newFollowedTopics = [...newMapFollowedTopics];
-       const newTopic = exhangeFollower(newFollowedTopics, willFollow, item.topic_id);
-       setFollowedTopic(
-         followedTopic.map((topic) => {
-           if (topic.topic_id === newTopic.topic_id) {
-             return newTopic;
-           }
-           return topic;
-         })
-       );
-     }
-     if (from === FROM_UNFOLLOWED_TOPIC) {
-       const newUnFollowedTopic = [...newMapUnfollowedTopics];
-       const newTopic = exhangeFollower(newUnFollowedTopic, willFollow, item.topic_id);
-       setUnfollowedTopic(
-         unfollowedTopic.map((topic) => {
-           if (topic.topic_id === newTopic.topic_id) {
-             return newTopic;
-           }
-           return topic;
-         })
-       );
-     }
-  };

+  const handleTopic = (from, willFollow, item, index) => {
+    if (from === FROM_FOLLOWED_TOPIC_INITIAL || from === FROM_UNFOLLOWED_TOPIC_INITIAL) {
+      updateTopicList(from, topics, willFollow, item);
+    } else if (from === FROM_FOLLOWED_TOPIC) {
+      updateTopicList(from, newMapFollowedTopics, willFollow, item);
+    } else if (from === FROM_UNFOLLOWED_TOPIC) {
+      updateTopicList(from, newMapUnfollowedTopics, willFollow, item);
+    }
+  };
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->